akka-testkit/src/main/scala/akka/testkit/TestKit.scala

```diff
@@ -625,12 +625,40 @@ trait TestKitBase {
 
   /**
    * Assert that no message is received for the specified time.
+   * NOTE! Supplied value is always dilated.
    */
-  def expectNoMsg(max: FiniteDuration) { expectNoMsg_internal(max.dilated) }
+  @deprecated(message = "Use expectNoMessage instead", since = "2.5.4")
+  def expectNoMsg(max: FiniteDuration) {
+    expectNoMsg_internal(max.dilated)
+  }
+
+  /**
+   * Assert that no message is received for the specified time.
+   */
+  def expectNoMessage(max: FiniteDuration) = expectNoMsg_internal(max)
 
   private def expectNoMsg_internal(max: FiniteDuration) {
-    val o = receiveOne(max)
-    assert(o eq null, s"received unexpected message $o")
+    val finish = System.nanoTime() + max.toNanos
+    val pollInterval = 100.millis
+
+    def leftNow = (finish - System.nanoTime()).nanos
+
+    var elem: AnyRef = queue.peekFirst()
+    var left = leftNow
+    while (left.toNanos > 0 && elem == null) {
+      //Use of (left / 2) gives geometric series limited by finish time similar to (1/2)^n limited by 1,
+      //so it is very precise
+      Thread.sleep(
+        pollInterval.toMillis min (left / 2).toMillis
+      )
+      left = leftNow
+      if (left.toNanos > 0) {
+        elem = queue.peekFirst()
+      }
+    }
+    val diff = (finish - left.toNanos).nanos
+    val m = s"received unexpected message $elem after ${diff.toCoarsest}"
```

Comment:
No it wouldn't have been OK with single sleep. 
I've tested it with single sleep and encountered the same problem as in #23224, because sleep is not precise and can last longer than specified time for 10 or more millis and that will lead to unchecked queue. This solution tends to  be much more precise. 
I've tried to get last queue check at latest possible moment before timeout.
About precision in the message... Do you think it is ok or too detailed info)))?